Refactor of lighting interfaces

shadielhajj opened the pull request:
Implementation of #211.

I think the lighting pipeline looks much cleaner and simpler now.

Let me know if it all looks good to you @patriciogonzalezvivo  and I'll port it to HLSL.

patriciogonzalezvivo commented:
oh Nice!! I will check it out tomorrow morning.

Thank you again!

patriciogonzalezvivo commented:
Super nice cleaning of all the redundant functions + new datatypes and helper functions !! <3

patriciogonzalezvivo commented on lighting/pbrLittle.glsl:
```diff
@@ -46,76 +46,60 @@ license:
+    float shadow = shadow(LIGHT_SHADOWMAP, vec2(LIGHT_SHADOWMAP_SIZE), (LIGHT_COORD).xy, (LIGHT_COORD).z);
     #elif defined(FNC_RAYMARCH_SOFTSHADOW)
-    _shadow *= raymarchSoftShadow(_position, L);
+    float shadow = raymarchSoftShadow(mat.position, shadingData.L);
```
A default needs to be added. Ex:

```glsl
#else
float shadow = 1.0;
#endif
```

patriciogonzalezvivo commented on lighting/gooch.glsl:
```diff
@@ -1,12 +1,11 @@
-// #include "specular.glsl"
-// #include "../sample/shadowPCF.glsl"
+#include "specular.glsl"
+#include "diffuse.glsl"
```
missing dependency
```glsl
#include "reflection.glsl"
```

shadielhajj replied:
Thanks for the catch. Fixed.

patriciogonzalezvivo commented on lighting/pbr.glsl:
```diff
@@ -45,65 +46,54 @@ license:
         LightPoint L = LightPointNew();
-        lightResolve(diffuseColor, specularColor, M, L, lightDiffuse, lightSpecular);
         #endif
+        lightResolve(L, mat, shadingData);
```
This fails if `LIGHT_DIRECTION` and  `LIGHT_POSITION` are not defined

suggestion
```glsl
        #if defined(LIGHT_DIRECTION)
        LightDirectional L = LightDirectionalNew();
        lightResolve(L, mat, shadingData);

        #elif defined(LIGHT_POSITION)
        LightPoint L = LightPointNew();
        lightResolve(L, mat, shadingData);
        #endif
```

shadielhajj replied:
Absolutely. Thanks.

patriciogonzalezvivo commented on lighting/light/pointEvaluate.glsl:
```diff
@@ -0,0 +1,60 @@
+#ifndef FNC_LIGHT_POINT
+#define FNC_LIGHT_POINT
+
+void lightPointEvaluate(LightPoint L, Material mat, ShadingData shadingData) {
```
`shadingData` doesn't return. Suggestion

```glsl

void lightPointEvaluate(LightPoint L, Material mat, inout ShadingData shadingData)
```

shadielhajj replied:
Good catch. Fixed.

patriciogonzalezvivo commented:
Getting the following visual regressions

![image](https://github.com/user-attachments/assets/57fe6832-aaa9-47e4-97ea-41975a4a96eb)

![image](https://github.com/user-attachments/assets/937a15a3-c3e7-4b21-950b-e111e353887c)

![image](https://github.com/user-attachments/assets/269ee96f-303f-4fe7-b6c4-dad1882b05b8)

![image](https://github.com/user-attachments/assets/b460cc90-a32e-43a7-b945-e90d451a1cbb)

![image](https://github.com/user-attachments/assets/f2e170fe-dbb0-4d76-80dc-e8fec845eb1e)

![image](https://github.com/user-attachments/assets/117292fb-9880-48de-adde-dac972e4130f)

![image](https://github.com/user-attachments/assets/d943c1ce-dc53-438a-88c0-d2086c2ce5e0)

![image](https://github.com/user-attachments/assets/081ac7b7-c783-4e90-9ba1-179266a1ab87)

shadielhajj commented:
> Getting the following visual regressions

On it. @patriciogonzalezvivo is there a script to get these nice side-by-side tests?

patriciogonzalezvivo commented:
I generate new screenshots locally and use vs code git tab to check on image that had change. I'm improving the pipeline of lygia_example to generate the makefile, readme and benchmark tests.

shadielhajj commented:
@patriciogonzalezvivo Should be all working now:

* `pbrLittle` examples: FIXED
* `pbrClearCoat`examples: FIXED
* AOV example. `lighting_raymarching_return.frag` needs to be updated to use the `RAYMARCH_AOV` define instead of the older `RAYMARCH_RETURN` see #206 
* `ssr` example this also needs modification on the client side to account for the new interface. In `lighting_ssr.frag` could to the following:`
```glsl
 #if defined(SCENE_BUFFER_0)
    // SPECULAR BUFFER
    ShadingData shadingData = shadingDataNew();
    shadingData.L = normalize(LIGHT_POSITION - (SURFACE_POSITION).xyz);
    shadingData.N = normalize(material.normal);
    shadingData.V = normalize(CAMERA_POSITION - (SURFACE_POSITION).xyz);
    shadingData.roughness =  material.roughness;
    shadingData.linearRoughness =  material.roughness;
    color += saturate(specular(shadingData));
    float spec =  saturate(.95 - material.roughness * 0.5);
    color.rgb *= 1.0-saturate( (powFast(spec, 4.) * 0.8 + 1.6 * (1.0-material.metallic)) );
    #else
``` 

Let me know if there's any visual regressions left, but I think we should be good now.

patriciogonzalezvivo commented:
There still some weird specular/reflections

![image](https://github.com/user-attachments/assets/27b656dc-c468-430e-95fc-920da0ebd1ca)

![image](https://github.com/user-attachments/assets/92e460c5-3cb3-400a-8fc1-3642ca1498ec)

![image](https://github.com/user-attachments/assets/0383c9d9-dc5f-4aec-afe9-6c3d9cfa9e0e)

and lack of depth information on raymarching

![image](https://github.com/user-attachments/assets/f5865053-272f-439f-b99f-d4394a5f985b)

patriciogonzalezvivo commented:
Oh, haven't notice that change. What `RAYMARCH_AOV` stands for? Could we leave `RAYMARCH_RETURN` instead?

shadielhajj commented:
@patriciogonzalezvivo AOV (Arbitrary Output Variables) is the name used in pro rendering packages (Unity, Redshift, Octane, ...), and is quite standard in the industry. But I can leave it as `RAYMARCH_RETURN ` if you prefer. Btw this change was done and documented in #206.

shadielhajj commented:
@patriciogonzalezvivo I think it's all fixed now! 😊
HLSL port should also be ready.
Take a look when you have a sec and please merge it if looks all good.

patriciogonzalezvivo commented:
![image](https://github.com/user-attachments/assets/10fef532-2b65-4b32-bc26-8caad161d3f9)

![image](https://github.com/user-attachments/assets/3eca32d6-0dbf-4cf3-b0a5-c956f1f32a1d)

I still see some differences here. The spheres on the on the back are a bit dull

Also:

![image](https://github.com/user-attachments/assets/aa990f1d-4cc5-4168-a554-8b0f648d885f)

shadielhajj commented:
@patriciogonzalezvivo Ah yes, yell spotted for the back sphere, I'll investigate.

Regarding the SSR example, as I mentioned in a previous comment, this one requires a client-side change (ie. a Patricio amend), since the interface of the `specular` function has changed. In `lighting_ssr.frag` you could do the following:
```glsl
 #if defined(SCENE_BUFFER_0)
    // SPECULAR BUFFER
    ShadingData shadingData = shadingDataNew();
    shadingData.L = normalize(LIGHT_POSITION - (SURFACE_POSITION).xyz);
    shadingData.N = normalize(material.normal);
    shadingData.V = normalize(CAMERA_POSITION - (SURFACE_POSITION).xyz);
    shadingData.roughness =  material.roughness;
    shadingData.linearRoughness =  material.roughness;
    color += saturate(specular(shadingData));
    float spec =  saturate(.95 - material.roughness * 0.5);
    color.rgb *= 1.0-saturate( (powFast(spec, 4.) * 0.8 + 1.6 * (1.0-material.metallic)) );
    #else
```

shadielhajj commented:
@patriciogonzalezvivo All the spheres look identical now in the clear coat examples. Hopefully, I didn't forget anything this time.
Thanks for your help testing the refactor, btw. It's hard to not miss anything on this scale.

patriciogonzalezvivo commented:
great work! thank you!

patriciogonzalezvivo commented:
We hit an important regression that manifest only on PLATFORM_WEBGL after this merge. 

```
';' : syntax error
```

![image](https://github.com/user-attachments/assets/d9b98983-27ce-49c0-a7a6-8d0511c7a72a)

Do you mind helping me to find it?

shadielhajj commented:
@patriciogonzalezvivo sure! How can I test this? I can't seem to be able to reproduce it with my toolset. And the error message is a bit vague.

shadielhajj commented:
@patriciogonzalezvivo One thing that comes to mind without being able to reproduce the error, are the defines in `specular`and `diffuse`. Might be worth trying different functions.

```glsl
#ifndef DIFFUSE_FNC 
#if defined(TARGET_MOBILE) || defined(PLATFORM_RPI) || defined(PLATFORM_WEBGL)
#define DIFFUSE_FNC diffuseLambert
#else
#define DIFFUSE_FNC diffuseOrenNayar
#endif  
#endif
``` 

```glsl
#ifndef SPECULAR_FNC
#if defined(TARGET_MOBILE) || defined(PLATFORM_RPI) || defined(PLATFORM_WEBGL)
#define SPECULAR_FNC specularBlinnPhongRoughness
#else
#define SPECULAR_FNC specularCookTorrance
#endif  
#endif
```

patriciogonzalezvivo commented:
Found it! Was pretty hard to spot and only manifesting on WebGL 1.0 : )

https://github.com/patriciogonzalezvivo/lygia/commit/556a1e31fd80f21406f370e4eaf7ca7a99c20362

Man... WebGL 1.0 is such a pain in the but.

Great work! That was a massive rearch!

patriciogonzalezvivo commented:
Question, finding the bug I notice the define in `raymarch.?lsl` is `#define ENG_RAYMARCH` it's ok if we revert this one to `#define FNC_RAYMARCH` for consistency?

shadielhajj commented:
@patriciogonzalezvivo Wow! Well spotted 😅😅😅 and sorry about that!
`FNC_RAYMARCH` is totally fine! `Not sure where this `ENG_` prefix comes from.
